fdbserver/DataDistribution.actor.cpp

```diff
@@ -1479,219 +1570,544 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+		} else if (audit->anyChildAuditFailed) {
+			audit->anyChildAuditFailed = false;
+			throw audit_storage_failed(); // retry
+		} else {
+			audit->coreState.setPhase(AuditPhase::Complete);
+		}
+		TraceEvent(SevVerbose, "DDAuditStorageCoreGotResult", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditState", audit->coreState.toString())
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		wait(persistAuditState(self->txnProcessor->context(),
+		                       audit->coreState,
+		                       "AuditStorageCore",
+		                       lockInfo,
+		                       self->context->isDDEnabled()));
+		TraceEvent(SevVerbose, "DDAuditStorageCoreSetResult", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditState", audit->coreState.toString())
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		removeAuditFromAuditMap(self,
+		                        audit->coreState.getType(),
+		                        audit->coreState.id,
+		                        audit->coreState.getPhase(),
+		                        audit->retryCount); // remove audit
+
+		TraceEvent(SevInfo, "DDAuditStorageCoreEnd", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditID", auditID)
+		    .detail("AuditType", auditType)
+		    .detail("Range", audit->coreState.range)
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+	} catch (Error& e) {
+		TraceEvent(SevDebug, "DDAuditStorageCoreError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("Context", context)
+		    .detail("AuditID", auditID)
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("AuditType", auditType)
+		    .detail("Range", audit->coreState.range)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		if (e.code() == error_code_actor_cancelled || e.code() == error_code_movekeys_conflict) {
+			throw e;
+		} else if (audit->retryCount < SERVER_KNOBS->AUDIT_RETRY_COUNT_MAX && e.code() != error_code_not_implemented) {
+			audit->retryCount++;
+			audit->actors.clear(true);
+			TraceEvent(SevVerbose, "DDAuditStorageCoreRetry", self->ddId)
+			    .detail("AuditID", auditID)
+			    .detail("AuditType", auditType)
+			    .detail("RetryCount", currentRetryCount)
+			    .detail("Contains", self->audits.contains(auditType) && self->audits[auditType].contains(auditID));
+			wait(delay(0.1));
+			TraceEvent(SevVerbose, "DDAuditStorageCoreRetryAfterWait", self->ddId)
+			    .detail("AuditID", auditID)
+			    .detail("AuditType", auditType)
+			    .detail("RetryCount", currentRetryCount)
+			    .detail("Contains", self->audits.contains(auditType) && self->audits[auditType].contains(auditID));
+			audit->setAuditRunActor(auditStorageCore(self, auditID, auditType, context, audit->retryCount));
```

Comment:
Do you think we can replace this block with `runAuditStorage()`? So that `runAuditStorage()` is the only entry point to start an audit?

Reply:
Yes. This is a more clean way. Thanks!